internal/controllers/compute/machine_ephemeralvolume_controller.go

```diff
@@ -0,0 +1,225 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compute
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	computev1alpha1 "github.com/onmetal/onmetal-api/api/compute/v1alpha1"
+	storagev1alpha1 "github.com/onmetal/onmetal-api/api/storage/v1alpha1"
+	computeclient "github.com/onmetal/onmetal-api/internal/client/compute"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+type MachineEphemeralVolumeReconciler struct {
+	client.Client
+}
+
+//+kubebuilder:rbac:groups=compute.api.onmetal.de,resources=machines,verbs=get;list;watch
+//+kubebuilder:rbac:groups=storage.api.onmetal.de,resources=volumes,verbs=get;list;watch;delete
+
+func (r *MachineEphemeralVolumeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	machine := &computev1alpha1.Machine{}
+	if err := r.Get(ctx, req.NamespacedName, machine); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.reconcileExists(ctx, log, machine)
+}
+
+func (r *MachineEphemeralVolumeReconciler) reconcileExists(ctx context.Context, log logr.Logger, machine *computev1alpha1.Machine) (ctrl.Result, error) {
+	if !machine.DeletionTimestamp.IsZero() {
+		log.V(1).Info("Machine is deleting, nothing to do")
+		return ctrl.Result{}, nil
+	}
+
+	return r.reconcile(ctx, log, machine)
+}
+
+func (r *MachineEphemeralVolumeReconciler) ephemeralMachineVolumeByName(machine *computev1alpha1.Machine) map[string]*storagev1alpha1.Volume {
+	res := make(map[string]*storagev1alpha1.Volume)
+	for _, machineVolume := range machine.Spec.Volumes {
+		ephemeral := machineVolume.Ephemeral
+		if ephemeral == nil {
+			continue
+		}
+
+		volumeName := computev1alpha1.MachineEphemeralVolumeName(machine.Name, machineVolume.Name)
+		volume := &storagev1alpha1.Volume{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:   machine.Namespace,
+				Name:        volumeName,
+				Labels:      ephemeral.VolumeTemplate.Labels,
+				Annotations: ephemeral.VolumeTemplate.Annotations,
+			},
+			Spec: ephemeral.VolumeTemplate.Spec,
+		}
+		_ = ctrl.SetControllerReference(machine, volume, r.Scheme())
```

Comment:
is there a reason for skipping the error check?

Reply:
same question for the other types

Reply:
The error can only occur if you're modifying an existing object (i.e. one you retrieved from the server). In this case since we're initializing the struct, we can be sure that there are no `OwnerReferences` and thus this cannot fail.